Applications/Spire/Source/Ui/ClosedFilterPanel.cpp

```diff
@@ -0,0 +1,216 @@
+      check_box->setLayoutDirection(Qt::RightToLeft);
+      check_box->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
+      return check_box;
+    });
+  set_style(*m_list_view, LIST_VIEW_STYLE(get_style(*m_list_view)));
+  for(auto i = 0; i < m_list_view->get_list_model()->get_size(); ++i) {
+    set_style(*m_list_view->get_list_item(i), LIST_ITEM_STYLE());
+  }
+  m_list_view->get_list_model()->connect_operation_signal(
+    std::bind_front(&ClosedFilterPanel::on_list_model_operation, this));
+  auto scrollable_list_box = new ScrollableListBox(*m_list_view);
+  scrollable_list_box->setMinimumSize(scale_width(160), scale_height(54));
+  scrollable_list_box->setMaximumHeight(scale_height(158));
+  if(scrollable_list_box->sizeHint().height() >
+      scrollable_list_box->maximumHeight()) {
+    scrollable_list_box->get_scroll_box().get_vertical_scroll_bar().show();
+  }
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(scrollable_list_box);
+  m_filter_panel = new FilterPanel(std::move(title), this, parent);
+  m_filter_panel->connect_reset_signal(
+    std::bind_front(&ClosedFilterPanel::on_reset, this));
+  window()->installEventFilter(this);
+}
+
+const std::shared_ptr<TableModel>& ClosedFilterPanel::get_model() const {
+  return m_model;
+}
+
+connection ClosedFilterPanel::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool ClosedFilterPanel::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::Close) {
+    m_filter_panel->hide();
+    hide();
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+bool ClosedFilterPanel::event(QEvent* event) {
+  if(event->type() == QEvent::ShowToParent) {
+    m_filter_panel->show();
+  } else if(event->type() == QEvent::HideToParent) {
+    m_filter_panel->hide();
+  }
+  return QWidget::event(event);
+}
+
+void ClosedFilterPanel::on_list_model_operation(
+    const ListModel::Operation& operation) {
+  visit(operation,
+    [=] (const ListModel::AddOperation& operation) {
+      set_style(*m_list_view->get_list_item(operation.m_index),
+        LIST_ITEM_STYLE());
+      invalidate_descendants(*window());
+      if(m_model->get<bool>(operation.m_index, 1)) {
+        m_submission->push(m_model->at(operation.m_index, 0));
+      }
+    },
+    [=] (const ListModel::RemoveOperation& operation) {
+      invalidate_descendants(*window());
+      auto index = m_submission->get_size();
+      while(--index >= 0) {
+        m_submission->remove(index);
+      }
+      for(auto i = 0; i < m_model->get_row_size(); ++i) {
+        if(m_model->get<bool>(i, 1)) {
+          m_submission->push(m_model->at(i, 0));
+        }
+      }
+    });
+}
+
+void ClosedFilterPanel::on_table_model_operation(
+    const TableModel::Operation& operation) {
+  visit(operation,
+    [=] (const TableModel::UpdateOperation& operation) {
```

Comment:
Capture by reference, you only capture by value when the closure outlives the lambda expression's lexical scope.